### PR TITLE
vim-patch:8.2.{0877, 0880, 0884, 0887, 0896} - port searchcount()

### PR DESCRIPTION
**Pull request by shadmansaleh**
https://github.com/vim/vim/commit/e8f5ec0d30b629d7166f0ad03434065d8bc822df

Adding searchcount function .

### Todo :
 - [x] Add related patches
 - [x] figure out why test are failing

### CODE REVIEW

**Comment by shadmansaleh** [comment]
### Test summery:
- linux tsan, osx, sourcehut failing beacause of luarocka . tsan and sourcehut passed on previous commit . osx failed because of luarocks on last commit too;
- MINGW-32 is failing on master and is irrelevent to this pr

**Comment by mhinz** [comment]
Please rebase so that only vim-patch commits are left.

And remove the gitignore commit. Temporary files can be cleaned with `git clean -fd`.

**Comment by shadmansaleh** [comment]
> Please rebase so that only vim-patch commits are left.

And remove the gitignore commit. Temporary files can be cleaned with git clean -fd.

Done .

File: src/nvim/testdir/test_search_stat.vim [hunk 1]
```diff
@@ -252,9 +329,9 @@ func Test_searchcount_in_statusline()
     function TestSearchCount() abort
       let search_count = searchcount()
       if !empty(search_count)
-	return '[' . search_count.current . '/' . search_count.total . ']'
+  return '[' . search_count.current . '/' . search_count.total . ']'
```
Comment (mhinz) [inline src/nvim/testdir/test_search_stat.vim:332]: This doesn't look correct. ;]

File: src/nvim/testdir/test_search_stat.vim [hunk 2]
```diff
@@ -252,9 +329,9 @@ func Test_searchcount_in_statusline()
...
       else
-	return ''
+  return ''
```
Comment (mhinz) [inline src/nvim/testdir/test_search_stat.vim:334]: Same

File: src/nvim/search.h [hunk 3]
```diff
@@ -49,6 +50,13 @@
 #define RE_BOTH         2       /* save pat in both patterns */
 #define RE_LAST         2       /* use last used pattern if "pat" is NULL */
 
+
```
Comment (mhinz) [inline src/nvim/search.h:53]: one blank line is enough

File: src/nvim/search.h [hunk 4]
```diff
@@ -49,6 +50,13 @@
...
+// Values for searchcount()
+#define SEARCH_STAT_DEF_TIMEOUT 40L
+#define SEARCH_STAT_DEF_MAX_COUNT 99
+#define SEARCH_STAT_BUF_LEN 12
+
+
```
Comment (mhinz) [inline src/nvim/search.h:59]: same

File: src/nvim/search.c [hunk 5]
```diff
@@ -4347,121 +4350,293 @@ int linewhite(linenr_T lnum)
...
+
+      // keep the message even after redraw, but don't put in history
+      msg_hist_off = true;
+      msg_ext_set_kind("search_count");
+      give_warning(msgbuf, false);
+      msg_hist_off = false;
+    }
+}
+
+
```
Comment (mhinz) [inline src/nvim/search.c:4415]: blank lines

File: src/nvim/search.c [hunk 6]
```diff
@@ -1002,7 +1002,7 @@ static int first_submatch(regmmatch_T *rp)
 /*
  * Highest level string search function.
  * Search for the 'count'th occurrence of pattern 'pat' in direction 'dirc'
- *		  If 'dirc' is 0: use previous dir.
+ *      If 'dirc' is 0: use previous dir.
```
Comment (mhinz) [inline src/nvim/search.c:1005]: spacing

**Comment by shadmansaleh** [comment]
Should I make these changes in the vim-patch commit or make a separate commit for fixes ?

File: src/nvim/search.c [hunk 7]
```diff
@@ -46,38 +46,38 @@
...
- * Two search patterns are remembered: One for the :substitute command and
- * one for other searches.  last_idx points to the one that was used the last
- * time.
- */
+//  This file contains various searching-related routines. These fall into
+//  three groups:
+//  1. string searches (for /, ?, n, and N)
+//  2. character searches within a single line (for f, F, t, T, etc)
+//  3. "other" kinds of searches like the '%' command, and 'word' searches.
+// /
```
Comment (mhinz) [inline src/nvim/search.c:59]: `//`

File: src/nvim/search.c [hunk 8]
```diff
@@ -46,38 +46,38 @@
...
+
+//  The actual string matching is done using a heavily modified version of
+//  Henry Spencer's regular expression library.  See regexp.c.
+
+
+
+// Two search patterns are remembered: One for the :substitute command and
+// one for other searches.  last_idx points to the one that was used the last
+// time.
+
```
Comment (mhinz) [inline src/nvim/search.c:80]: This whole thing above is supposed to be one block, so if we're at it anyway, remove all multiple blank lines and add `//` for blank lines.

```c
// I'm the best block comment.
//
// Indeed.
```

File: src/nvim/testdir/test_search_stat.vim (hunk 1 above)
Comment (shadmansaleh) [reply src/nvim/testdir/test_search_stat.vim]: fixed

**Comment by shadmansaleh** [comment]
@mhinz I've fixed the issues you found :)

**Comment by shadmansaleh** [comment]
@janlazo any chance this can be merged in 0.5 ?

File: runtime/doc/eval.txt [hunk 9]
```diff
@@ -7292,6 +7292,126 @@ search({pattern} [, {flags} [, {stopline} [, {timeout}]]])	*search()*
 		The 'n' flag tells the function not to move the cursor.
 
 
+searchcount([{options}])					*searchcount()*
```
Comment (janlazo) [inline runtime/doc/eval.txt:7295]: `searchcount([{options}])	Dict	get or update search stats`
is missing in the `Builtin Functions` section.

Comment (shadmansaleh) [reply runtime/doc/eval.txt:7288]: Nice catch . Added :)

File: src/nvim/search.c [hunk 10]
```diff
@@ -1383,11 +1381,14 @@ int do_search(
         && c != FAIL
         && !shortmess(SHM_SEARCHCOUNT)
         && msgbuf != NULL) {
-      search_stat(dirc, &pos, show_top_bot_msg, msgbuf,
```
Comment (janlazo) [inline src/nvim/search.c:1386]: `#define SEARCH_STAT_BUF_LEN 12` wasn't deleted on line 1045.

File: src/nvim/search.c [hunk 11]
```diff
@@ -4347,121 +4348,288 @@ int linewhite(linenr_T lnum)
 }
 
 // Add the search count "[3/19]" to "msgbuf".
-// When "recompute" is true Always recompute the numbers.
-static void search_stat(int dirc, pos_T *pos,
-                        bool show_top_bot_msg, char_u *msgbuf, bool recompute)
+// See update_search_stat() for other arguments.
+static void cmdline_search_stat(int dirc, pos_T *pos, pos_T *cursor_pos,
+                                int show_top_bot_msg, char_u  *msgbuf,
```
Comment (janlazo) [inline src/nvim/search.c:4353]: `bool show_top_bot_msg` is correct.

File: src/nvim/search.c [hunk 12]
```diff
@@ -4347,121 +4348,288 @@ int linewhite(linenr_T lnum)
...
+                                int recompute, int maxcount, long timeout)
```
Comment (janlazo) [inline src/nvim/search.c:4354]: `bool recompute` is correct.

In general, don't change `bool` (or TriState) to `int` unless int values are used.
If int values are used, then always use int values even if Vim uses `FALSE` and `TRUE` for int parameters.

File: src/nvim/search.c (hunk 10 above)
Comment (shadmansaleh) [reply src/nvim/search.c:1386]: Oops forgot it was moved to `search.h` . Deleted .

File: src/nvim/search.c [hunk 13]
```diff
@@ -4347,121 +4347,288 @@ int linewhite(linenr_T lnum)
...
+        } else if (stat.cnt > maxcount) {
+          vim_snprintf(t, SEARCH_STAT_BUF_LEN, "[%d/>%d]",
+                       stat.cur, maxcount);
+        } else {
+          vim_snprintf(t, SEARCH_STAT_BUF_LEN, "[%d/%d]",
+                       stat.cur, stat.cnt);
+        }
+      }
+
+      len = STRLEN(t);
```
Comment (janlazo) [inline src/nvim/search.c:4391]: `size_t len = strlen(t);` because `t` is `char` type.

File: src/nvim/search.c (hunk 11 above)
Comment (shadmansaleh) [reply src/nvim/search.c]: updated

File: src/nvim/search.c (hunk 12 above)
Comment (shadmansaleh) [reply src/nvim/search.c]: changed . + recompute in `update_search_state` changed to `bool`

File: src/nvim/search.c [hunk 14]
```diff
@@ -4347,121 +4347,288 @@ int linewhite(linenr_T lnum)
...
+                               searchstat_T *stat, int recompute, int maxcount,
+                               long timeout)
+{
+    int             save_ws = p_ws;
+    int             wraparound = false;
+    pos_T           p = (*pos);
+    static pos_T    lastpos = { 0, 0, 0 };
     static int      cur = 0;
     static int      cnt = 0;
+    static int      exact_match = false;
```
Comment (janlazo) [inline src/nvim/search.c:4428]: It's documented as a boolean.
`exact_match` uses true/false values only.
Why is this int?

File: src/nvim/search.c [hunk 15]
```diff
@@ -4347,121 +4347,288 @@ int linewhite(linenr_T lnum)
...
     }
 
-    if (equalpos(lastpos, curwin->w_cursor) && !wraparound
-        && (dirc == '/' ? cur < cnt : cur > 0)) {
-      cur += dirc == '/' ? 1 : -1;
+    if (equalpos(lastpos, *cursor_pos) && !wraparound
+        && (dirc == 0 || dirc == '/' ? cur < cnt : cur > 0)) {
+      cur += dirc == 0 ? 0 : dirc == '/' ? 1 : -1;
     } else {
+      int done_search = false;
```
Comment (janlazo) [inline src/nvim/search.c:4474]: bool

**Review by janlazo — CHANGES_REQUESTED** [review]

File: src/nvim/search.c (hunk 13 above)
Comment (shadmansaleh) [reply src/nvim/search.c:4391]: But len is alrwady declared as size_t at line 4361

File: src/nvim/search.c (hunk 14 above)
Comment (shadmansaleh) [reply src/nvim/search.c]: because I'm blind 🤦‍♂️

File: src/nvim/search.c (hunk 15 above)
Comment (shadmansaleh) [reply src/nvim/search.c]: I've changed this and few more to bool . See diff of last commit

**Review by janlazo — CHANGES_REQUESTED** [review]

File: src/nvim/search.c [hunk 16]
```diff
@@ -4347,121 +4347,288 @@ int linewhite(linenr_T lnum)
...
-
-      len = STRLEN(t);
-      if (show_top_bot_msg && len + 2 < SEARCH_STAT_BUF_LEN) {
-        memmove(t + 2, t, len);
-        t[0] = 'W';
-        t[1] = ' ';
-        len += 2;
+      di = tv_dict_find(dict, (const char *)"recompute", -1);
+      if (di != NULL) {
+        recompute = tv_get_number_chk(&di->di_tv, &error);
```
[the remark this answers is not in the document]
Comment (shadmansaleh) [reply src/nvim/search.c:4564]: I think that patch isn't in this pr. I couldn't find replacement of dict_get_number and that patch was about running test in vim9 so I didn't port it

File: src/nvim/search.c (hunk 13 above)
Comment (shadmansaleh) [reply src/nvim/search.c]: Done

File: src/nvim/search.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (shadmansaleh) [reply src/nvim/search.c:4563]: `tv_dict_get_bool` doesn't exist ?

[the remark this answers is not in the document]
Comment (janlazo) [reply src/nvim/search.c:4563]: not yet.
Just taking note of future changes.

**Review by janlazo — APPROVED** [review]
Good enough.

Avoid whitespace,alignment changes unless the vim patch modifies the line because the reviewer has to go through lines. Style changes can go to a separate commit.

**Comment by shadmansaleh** [comment]
> Avoid whitespace,alignment changes unless the vim patch modifies the line because the reviewer has to go through lines. Style changes can go to a separate commit.

Ok . Thanks will remember that from the next time :)

**Pull request merged**